google/cloud/pubsub/internal/subscriber_tracing_connection.cc

```diff
@@ -0,0 +1,97 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/pubsub/internal/subscriber_tracing_connection.h"
+#include "google/cloud/pubsub/options.h"
+#include "google/cloud/grpc_options.h"
+#include "google/cloud/internal/opentelemetry.h"
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include "opentelemetry/trace/semantic_conventions.h"
+#include "opentelemetry/trace/span_startoptions.h"
+#endif  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+namespace google {
+namespace cloud {
+namespace pubsub_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+namespace {
+
+opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> StartPullSpan() {
+  auto const& current = internal::CurrentOptions();
+  auto const& subscription = current.get<pubsub::SubscriptionOption>();
+  namespace sc = opentelemetry::trace::SemanticConventions;
+  opentelemetry::trace::StartSpanOptions options;
+  options.kind = opentelemetry::trace::SpanKind::kConsumer;
+  auto span = internal::MakeSpan(
+      subscription.subscription_id() + " receive",
+      {{sc::kMessagingSystem, "gcp_pubsub"},
+       {sc::kMessagingOperation, "receive"},
+       {sc::kCodeFunction, "pubsub::SubscriberConnection::Pull"},
+       {sc::kMessagingDestinationTemplate, subscription.FullName()}},
+      options);
+  return span;
+}
+
+}  // namespace
+
+class SubscriberTracingConnection : public pubsub::SubscriberConnection {
```

Comment:
It looks like this could be in the anonymous namespace.

Reply:
Good catch. Done.